app/controllers/application_controller.rb

```diff
@@ -717,12 +717,21 @@ def build_not_found_page(status = 404, layout = false)
       layout = 'application' if layout == 'no_ember'
     end
 
-    category_topic_ids = Category.pluck(:topic_id).compact
+    if !SiteSetting.login_required? || current_user
+      key = "page_not_found_topics"
+      if @topics_partial = $redis.get(key)
+        @topics_partial = @topics_partial.html_safe
+      else
+        category_topic_ids = Category.pluck(:topic_id).compact
+        @top_viewed = TopicQuery.new(nil, except_topic_ids: category_topic_ids).list_top_for("monthly").topics.first(10)
+        @recent = Topic.includes(:category).where.not(id: category_topic_ids).recent(10)
+        @topics_partial = render_to_string partial: '/exceptions/not_found_topics'
```

Comment:
Does `render_to_string` call `html_safe`?

Reply:
`render_to_string` builds an `ActionView::OutputBuffer` which inherits `ActiveSupport::SafeBuffer` and `html_safe` constructs a new `ActiveSupport::SafeBuffer`.

In other words, `render_to_string` does not call `html_safe`, but they have a common ancestor.